apollo-federation/src/operation/mod.rs

```diff
@@ -201,7 +205,79 @@ pub(crate) struct NormalizedDefer {
     pub operation: Operation,
     pub has_defers: bool,
     pub assigned_defer_labels: IndexSet<String>,
-    pub defer_conditions: IndexMap<String, IndexSet<String>>,
+    pub defer_conditions: IndexMap<Name, IndexSet<String>>,
+}
+
+struct DeferNormalizer {
+    used_labels: HashSet<String>,
+    assigned_labels: IndexSet<String>,
+    problems: HashSet<SelectionKey>,
+    conditions: IndexMap<Name, IndexSet<String>>,
+    label_offset: usize,
+}
+
+impl DeferNormalizer {
+    fn new(selection_set: &SelectionSet) -> Result<Self, FederationError> {
+        let mut digest = Self {
+            used_labels: HashSet::default(),
+            problems: HashSet::default(),
+            label_offset: 0,
+            assigned_labels: IndexSet::default(),
+            conditions: IndexMap::default(),
+        };
+        let mut stack = selection_set
+            .into_iter()
+            .map(|(_, sel)| sel)
+            .collect::<Vec<_>>();
+        while let Some(selection) = stack.pop() {
+            // TODO: Does this need to be checking FragmentSpread + InlineFragment or just
```

Comment:
when this is initialized all fragments are already inlined/expanded so we shouldn't have any `FragmentSpread`s